# ISA-PHM JSON Analysis

Analysis of ISA-JSON format file: `isa-phm-out (35).json`

This notebook displays protocol parameters (measurement and processing) with their values and units across all sensors.

## 1. Import Libraries

In [16]:
import json
import pandas as pd

print("✅ Libraries imported successfully")

✅ Libraries imported successfully


## 2. Load Data Files

In [17]:
# Load ISA-JSON file
with open("isa-phm-out (35).json", "r") as file:
    data = json.load(file)

print(f"✅ Loaded ISA-JSON: {len(data['studies'])} studies")

✅ Loaded ISA-JSON: 16 studies


In [18]:
# Check for units (ISA standard uses 'unitCategories' at study level)
study1 = data['studies'][0]
units_count = len(study1.get('unitCategories', []))
print(f"Study 1 has {units_count} unit categories")

if units_count == 0:
    print("\n⚠️  WARNING: No units found in ISA-JSON!")
    print("   You may need to regenerate with the fixed conversion script.")
else:
    print(f"\n✅ Units present! All {units_count} unit categories:")
    for i, unit in enumerate(study1['unitCategories'], 1):
        print(f"   {i}. {unit.get('annotationValue', 'N/A')}")

Study 1 has 10 unit categories

✅ Units present! All 10 unit categories:
   1. mm
   2. rev/min
   3. mm/rev
   4. kHz
   5. V
   6. mV/g
   7. dB/octave
   8. Hz
   9. ms
   10. MHz


## 3. Basic Structure Analysis

Quick overview of the ISA-JSON structure: studies, runs, and assays.

In [19]:
# Number of studies
print(f"Total studies: {len(data['studies'])}")

Total studies: 16


In [20]:
# Study overview
for study in data["studies"]:
    print(f"Study ID: {study['identifier']}")
    print(f"Title: {study['title']}")
    print(f"Description: {study['description']}")
    print(f"Number of assays: {len(study['assays'])}")
    print(f"Number of runs: {next((comment.get('value') for comment in study.get('comments', []) if comment.get('name') == 'total_runs'), 'n/a')}")
    print("-" * 40)

Study ID: 4d938dab-b036-48e8-a781-1b3b80680d50
Title: Case 1
Description: Combination 1
Number of assays: 6
Number of runs: 17
----------------------------------------
Study ID: dd048b03-c052-445d-9f0c-84884e49abd5
Title: Case 2
Description: Combination 2
Number of assays: 6
Number of runs: 14
----------------------------------------
Study ID: e13e4dbe-059d-44f8-9f4c-dc07edbfed6d
Title: Case 3
Description: Combination 3
Number of assays: 6
Number of runs: 14
----------------------------------------
Study ID: 086a437c-6728-4b08-bcc5-4035074b67a5
Title: Case 4
Description: Combination 4
Number of assays: 6
Number of runs: 7
----------------------------------------
Study ID: 8ab8ae27-bcfa-4487-a5b5-13bb8b484e7f
Title: Case 5
Description: Combination 5
Number of assays: 6
Number of runs: 6
----------------------------------------
Study ID: 6f55b561-f79a-4f0c-acd1-b3d8ffba6d29
Title: Case 6
Description: Combination 6
Number of assays: 6
Number of runs: 1
------------------------------------

In [21]:
# Assay details for each study
for study in data["studies"]:
    print(f"Study ID: {study['identifier']}")
    print("=" * 40)
    for assay in study["assays"]:
        print(f"Assay Name: {assay['filename']}")
        print(f"Measurement Type: {assay['measurementType']['annotationValue']}")
        print(f"Technology Type: {assay['technologyType']['annotationValue']}")
        print(f"Platform: {assay['technologyPlatform']}")
        print(f"Number of data files: {len(assay['dataFiles'])}")
        print(f"Number of runs: {len(assay['materials']['samples'])}")
        print("-" * 40)
    print()

Study ID: 4d938dab-b036-48e8-a781-1b3b80680d50
Assay Name: a_st01_se01
Measurement Type: Vibration
Technology Type: Accelerometer
Platform: Wilcoxon 786B-10
Number of data files: 34
Number of runs: 17
----------------------------------------
Assay Name: a_st01_se02
Measurement Type: Vibration
Technology Type: Accelerometer
Platform: Wilcoxon 786B-10
Number of data files: 34
Number of runs: 17
----------------------------------------
Assay Name: a_st01_se03
Measurement Type: Current
Technology Type: Current Clamp
Platform: CTA 213
Number of data files: 34
Number of runs: 17
----------------------------------------
Assay Name: a_st01_se04
Measurement Type: Current
Technology Type: Current Clamp
Platform: CTA 213
Number of data files: 34
Number of runs: 17
----------------------------------------
Assay Name: a_st01_se05
Measurement Type: Acoustic Emission
Technology Type: Microphone
Platform: WD 975
Number of data files: 34
Number of runs: 17
----------------------------------------
Assay

In [22]:
# Quick overview: count key elements per study
for i, study in enumerate(data["studies"], 1):
    print(f"Study {i}: {study['title']}")
    print(f"  Assays: {len(study.get('assays', []))}")
    print(f"  Protocols: {len(study.get('protocols', []))}")
    print(f"  Factors: {len(study.get('factors', []))}")
    print(f"  Samples: {len(study.get('materials', {}).get('samples', []))}")
    print(f"  Characteristic Categories: {len(study.get('characteristicCategories', []))}")
    print()

Study 1: Case 1
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 17
  Characteristic Categories: 4

Study 2: Case 2
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 14
  Characteristic Categories: 4

Study 3: Case 3
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 14
  Characteristic Categories: 4

Study 4: Case 4
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 7
  Characteristic Categories: 4

Study 5: Case 5
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 6
  Characteristic Categories: 4

Study 6: Case 6
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 1
  Characteristic Categories: 4

Study 7: Case 7
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 8
  Characteristic Categories: 4

Study 8: Case 8
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 6
  Characteristic Categories: 4

Study 9: Case 9
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 9
  Characteristic Categories: 4

Study 10: Case 10
  Assays: 6
  Protocols: 13
  Factors: 6
  Samples: 10
  Char

In [23]:
# Detailed check: count data files and process sequences per assay
for i, study in enumerate(data["studies"], 1):
    print(f"Study {i}: {study['title']}")
    for j, assay in enumerate(study.get('assays', []), 1):
        data_files = len(assay.get('dataFiles', []))
        processes = len(assay.get('processSequence', []))
        print(f"  Assay {j}: {data_files} data files, {processes} processes")
    if not study.get('assays'):
        print("  ⚠️ NO ASSAYS")
    print()

Study 1: Case 1
  Assay 1: 34 data files, 34 processes
  Assay 2: 34 data files, 34 processes
  Assay 3: 34 data files, 34 processes
  Assay 4: 34 data files, 34 processes
  Assay 5: 34 data files, 34 processes
  Assay 6: 34 data files, 34 processes

Study 2: Case 2
  Assay 1: 28 data files, 28 processes
  Assay 2: 28 data files, 28 processes
  Assay 3: 28 data files, 28 processes
  Assay 4: 28 data files, 28 processes
  Assay 5: 28 data files, 28 processes
  Assay 6: 28 data files, 28 processes

Study 3: Case 3
  Assay 1: 28 data files, 28 processes
  Assay 2: 28 data files, 28 processes
  Assay 3: 28 data files, 28 processes
  Assay 4: 28 data files, 28 processes
  Assay 5: 28 data files, 28 processes
  Assay 6: 28 data files, 28 processes

Study 4: Case 4
  Assay 1: 14 data files, 14 processes
  Assay 2: 14 data files, 14 processes
  Assay 3: 14 data files, 14 processes
  Assay 4: 14 data files, 14 processes
  Assay 5: 14 data files, 14 processes
  Assay 6: 14 data files, 14 process

In [24]:
# Check if protocols have parameters (empty protocols are a sign of incomplete data)
for i, study in enumerate(data["studies"], 1):
    print(f"Study {i}: {study['title']}")
    for protocol in study.get('protocols', []):
        params = len(protocol.get('parameters', []))
        print(f"  Protocol '{protocol.get('name', 'unnamed')}': {params} parameters")
    if not study.get('protocols'):
        print("  ⚠️ NO PROTOCOLS")
    print()

Study 1: Case 1
  Protocol 'Experiment Preparation Protocol': 0 parameters
  Protocol 'Vibration measurement (2328ca4d-4fd3-4c7c-a82d-9a518f899b52)': 14 parameters
  Protocol 'Vibration measurement (78243c46-abc2-46a4-a17f-ec3af1fc749c)': 14 parameters
  Protocol 'Current measurement (91ac3682-0815-4121-88be-dfbbdeb51aad)': 14 parameters
  Protocol 'Current measurement (ab124d09-2206-4f29-b4c8-d399f826c83e)': 14 parameters
  Protocol 'Acoustic Emission measurement (09f9a5e2-53f7-4983-8f2f-539dff31c508)': 14 parameters
  Protocol 'Acoustic Emission measurement (725e0df6-ee35-4596-b2b5-12cc31a18e7d)': 14 parameters
  Protocol 'Vibration processing (2328ca4d-4fd3-4c7c-a82d-9a518f899b52)': 7 parameters
  Protocol 'Vibration processing (78243c46-abc2-46a4-a17f-ec3af1fc749c)': 7 parameters
  Protocol 'Current processing (91ac3682-0815-4121-88be-dfbbdeb51aad)': 0 parameters
  Protocol 'Current processing (ab124d09-2206-4f29-b4c8-d399f826c83e)': 0 parameters
  Protocol 'Acoustic Emission proce

In [25]:
# Compare first complete study vs last incomplete study
first_study = data["studies"][0]  # Study 1
last_study = data["studies"][-1]  # Study 16

print("=" * 60)
print("STUDY 1 (Should be complete):")
print("=" * 60)
print(f"Assays: {len(first_study.get('assays', []))}")
print(f"Protocols: {len(first_study.get('protocols', []))}")
print(f"Factors: {len(first_study.get('factors', []))}")
print(f"Samples: {len(first_study.get('materials', {}).get('samples', []))}")
print(f"People: {len(first_study.get('people', []))}")

print("\n" + "=" * 60)
print("STUDY 16 (Should be incomplete):")
print("=" * 60)
print(f"Assays: {len(last_study.get('assays', []))}")
print(f"Protocols: {len(last_study.get('protocols', []))}")
print(f"Factors: {len(last_study.get('factors', []))}")
print(f"Samples: {len(last_study.get('materials', {}).get('samples', []))}")
print(f"People: {len(last_study.get('people', []))}")

STUDY 1 (Should be complete):
Assays: 6
Protocols: 13
Factors: 6
Samples: 17
People: 2

STUDY 16 (Should be incomplete):
Assays: 6
Protocols: 13
Factors: 6
Samples: 6
People: 2


In [26]:
# Summary table showing completeness
import pandas as pd

summary_data = []
for i, study in enumerate(data["studies"], 1):
    summary_data.append({
        "Study": i,
        "Title": study['title'],
        "Assays": len(study.get('assays', [])),
        "Protocols": len(study.get('protocols', [])),
        "Factors": len(study.get('factors', [])),
        "Samples": len(study.get('materials', {}).get('samples', [])),
        "People": len(study.get('people', [])),
        "Complete": "✅" if len(study.get('assays', [])) > 0 and len(study.get('protocols', [])) > 0 else "❌"
    })

df = pd.DataFrame(summary_data)
print(df.to_string(index=False))

 Study   Title  Assays  Protocols  Factors  Samples  People Complete
     1  Case 1       6         13        6       17       2        ✅
     2  Case 2       6         13        6       14       2        ✅
     3  Case 3       6         13        6       14       2        ✅
     4  Case 4       6         13        6        7       2        ✅
     5  Case 5       6         13        6        6       2        ✅
     6  Case 6       6         13        6        1       2        ✅
     7  Case 7       6         13        6        8       2        ✅
     8  Case 8       6         13        6        6       2        ✅
     9  Case 9       6         13        6        9       2        ✅
    10 Case 10       6         13        6       10       2        ✅
    11 Case 11       6         13        6       23       2        ✅
    12 Case 12       6         13        6       15       2        ✅
    13 Case 13       6         13        6       15       2        ✅
    14 Case 14       6         13 

## Analyze Process Sequences, Inputs/Outputs, and Parameters

In [27]:
# Check process structure: inputs, outputs, and parameter values
for i, study in enumerate(data["studies"][:3], 1):  # Check first 3 studies in detail
    print(f"\n{'='*70}")
    print(f"STUDY {i}: {study['title']}")
    print(f"{'='*70}")
    
    for assay in study.get('assays', [])[:1]:  # Check first assay of each study
        print(f"\nAssay: {assay.get('filename', 'unknown')}")
        print(f"{'—'*70}")
        
        for j, process in enumerate(assay.get('processSequence', [])[:5], 1):  # First 5 processes
            print(f"\n  Process {j}:")
            print(f"    Protocol: {process.get('executesProtocol', {}).get('name', 'unknown')}")
            print(f"    Inputs: {len(process.get('inputs', []))}")
            print(f"    Outputs: {len(process.get('outputs', []))}")
            print(f"    Parameter Values: {len(process.get('parameterValues', []))}")
            
            # Show parameter details
            for param_val in process.get('parameterValues', [])[:3]:  # First 3 params
                category = param_val.get('category', {})
                param_name = category.get('parameterName', {}).get('annotationValue', 'unknown')
                value = param_val.get('value', 'N/A')
                unit = param_val.get('unit', {})
                unit_term = unit.get('annotationValue', '') if unit else ''
                print(f"      • {param_name}: {value} {unit_term}".strip())


STUDY 1: Case 1

Assay: a_st01_se01
——————————————————————————————————————————————————————————————————————

  Process 1:
    Protocol: unknown
    Inputs: 1
    Outputs: 1
    Parameter Values: 9
• unknown: Table
• unknown: 13
• unknown: High pass

  Process 2:
    Protocol: unknown
    Inputs: 1
    Outputs: 1
    Parameter Values: 7
• unknown: ITHACO 4302
  bandpass
• unknown: 24
• unknown: 400

  Process 3:
    Protocol: unknown
    Inputs: 1
    Outputs: 1
    Parameter Values: 9
• unknown: Table
• unknown: 13
• unknown: High pass

  Process 4:
    Protocol: unknown
    Inputs: 1
    Outputs: 1
    Parameter Values: 7
• unknown: ITHACO 4302
  bandpass
• unknown: 24
• unknown: 400

  Process 5:
    Protocol: unknown
    Inputs: 1
    Outputs: 1
    Parameter Values: 9
• unknown: Table
• unknown: 13
• unknown: High pass

STUDY 2: Case 2

Assay: a_st02_se01
——————————————————————————————————————————————————————————————————————

  Process 1:
    Protocol: unknown
    Inputs: 1
    Out

In [28]:
# Count parameter values across all processes in each study
for i, study in enumerate(data["studies"], 1):
    print(f"Study {i}: {study['title']}")
    
    total_processes = 0
    total_params = 0
    processes_with_params = 0
    processes_without_params = 0
    
    for assay in study.get('assays', []):
        for process in assay.get('processSequence', []):
            total_processes += 1
            param_count = len(process.get('parameterValues', []))
            total_params += param_count
            
            if param_count > 0:
                processes_with_params += 1
            else:
                processes_without_params += 1
    
    print(f"  Total processes: {total_processes}")
    print(f"  Processes with parameters: {processes_with_params}")
    print(f"  Processes without parameters: {processes_without_params}")
    print(f"  Total parameter values: {total_params}")
    print(f"  Average params per process: {total_params/total_processes if total_processes > 0 else 0:.2f}")
    print()

Study 1: Case 1
  Total processes: 204
  Processes with parameters: 170
  Processes without parameters: 34
  Total parameter values: 1054
  Average params per process: 5.17

Study 2: Case 2
  Total processes: 168
  Processes with parameters: 140
  Processes without parameters: 28
  Total parameter values: 868
  Average params per process: 5.17

Study 3: Case 3
  Total processes: 168
  Processes with parameters: 140
  Processes without parameters: 28
  Total parameter values: 868
  Average params per process: 5.17

Study 4: Case 4
  Total processes: 84
  Processes with parameters: 70
  Processes without parameters: 14
  Total parameter values: 434
  Average params per process: 5.17

Study 5: Case 5
  Total processes: 72
  Processes with parameters: 60
  Processes without parameters: 12
  Total parameter values: 372
  Average params per process: 5.17

Study 6: Case 6
  Total processes: 12
  Processes with parameters: 10
  Processes without parameters: 2
  Total parameter values: 62
  Ave

In [29]:
# Check input/output file references in processes
for i, study in enumerate(data["studies"][:3], 1):  # First 3 studies
    print(f"\n{'='*70}")
    print(f"STUDY {i}: {study['title']}")
    print(f"{'='*70}")
    
    for j, assay in enumerate(study.get('assays', [])[:2], 1):  # First 2 assays
        print(f"\nAssay {j}: {assay.get('filename', 'unknown')}")
        
        # Check process chain
        for k, process in enumerate(assay.get('processSequence', [])[:3], 1):  # First 3 processes
            inputs = process.get('inputs', [])
            outputs = process.get('outputs', [])
            
            print(f"  Process {k}:")
            if inputs:
                input_type = inputs[0].get('@id', '').split('/')[0]
                print(f"    Input: {input_type} ({len(inputs)} items)")
            if outputs:
                output_type = outputs[0].get('@id', '').split('/')[0]
                print(f"    Output: {output_type} ({len(outputs)} items)")
            
            # Check if process has next/previous links
            has_next = 'nextProcess' in process
            has_prev = 'previousProcess' in process
            print(f"    Linked: {'next' if has_next else ''}{' & ' if has_next and has_prev else ''}{'prev' if has_prev else ''}")


STUDY 1: Case 1

Assay 1: a_st01_se01
  Process 1:
    Input: #sample (1 items)
    Output: #data_file (1 items)
    Linked: next
  Process 2:
    Input: #data_file (1 items)
    Output: #data_file (1 items)
    Linked: next & prev
  Process 3:
    Input: #sample (1 items)
    Output: #data_file (1 items)
    Linked: next & prev

Assay 2: a_st01_se02
  Process 1:
    Input: #sample (1 items)
    Output: #data_file (1 items)
    Linked: next
  Process 2:
    Input: #data_file (1 items)
    Output: #data_file (1 items)
    Linked: next & prev
  Process 3:
    Input: #sample (1 items)
    Output: #data_file (1 items)
    Linked: next & prev

STUDY 2: Case 2

Assay 1: a_st02_se01
  Process 1:
    Input: #sample (1 items)
    Output: #data_file (1 items)
    Linked: next
  Process 2:
    Input: #data_file (1 items)
    Output: #data_file (1 items)
    Linked: next & prev
  Process 3:
    Input: #sample (1 items)
    Output: #data_file (1 items)
    Linked: next & prev

Assay 2: a_st02_se02

In [30]:
# Verify data file connections: are all data files referenced in processes?
for i, study in enumerate(data["studies"][:5], 1):  # First 5 studies
    print(f"\nStudy {i}: {study['title']}")
    
    for j, assay in enumerate(study.get('assays', [])[:1], 1):  # First assay
        data_files = assay.get('dataFiles', [])
        process_outputs = set()
        
        # Collect all output file IDs from processes
        for process in assay.get('processSequence', []):
            for output in process.get('outputs', []):
                process_outputs.add(output.get('@id', ''))
        
        # Check if all data files are referenced
        unreferenced = 0
        for df in data_files:
            if df.get('@id') not in process_outputs:
                unreferenced += 1
        
        print(f"  Assay {assay.get('filename', 'unknown')}:")
        print(f"    Total data files: {len(data_files)}")
        print(f"    Referenced in processes: {len(process_outputs)}")
        print(f"    Unreferenced: {unreferenced}")
        print(f"    Status: {'✅ All connected' if unreferenced == 0 else '⚠️ Some disconnected'}")


Study 1: Case 1
  Assay a_st01_se01:
    Total data files: 34
    Referenced in processes: 34
    Unreferenced: 0
    Status: ✅ All connected

Study 2: Case 2
  Assay a_st02_se01:
    Total data files: 28
    Referenced in processes: 28
    Unreferenced: 0
    Status: ✅ All connected

Study 3: Case 3
  Assay a_st03_se01:
    Total data files: 28
    Referenced in processes: 28
    Unreferenced: 0
    Status: ✅ All connected

Study 4: Case 4
  Assay a_st04_se01:
    Total data files: 14
    Referenced in processes: 14
    Unreferenced: 0
    Status: ✅ All connected

Study 5: Case 5
  Assay a_st05_se01:
    Total data files: 12
    Referenced in processes: 12
    Unreferenced: 0
    Status: ✅ All connected


In [31]:
# Compare Study 1 vs Study 8 process details (to see if Study 8 is really complete)
studies_to_compare = [0, 7, 15]  # Study 1, 8, and 16 (0-indexed)

for idx in studies_to_compare:
    study = data["studies"][idx]
    print(f"\n{'='*70}")
    print(f"STUDY {idx+1}: {study['title']}")
    print(f"{'='*70}")
    
    # Take first assay as sample
    if study.get('assays'):
        assay = study['assays'][0]
        processes = assay.get('processSequence', [])
        
        print(f"Assay: {assay.get('filename', 'unknown')}")
        print(f"Total processes: {len(processes)}")
        
        if processes:
            # Check first process
            first_process = processes[0]
            print(f"\nFirst Process:")
            print(f"  Protocol: {first_process.get('executesProtocol', {}).get('name', 'unknown')}")
            print(f"  Parameters: {len(first_process.get('parameterValues', []))}")
            
            # Sample parameter
            params = first_process.get('parameterValues', [])
            if params:
                param = params[0]
                category_name = param.get('category', {}).get('parameterName', {}).get('annotationValue', 'N/A')
                print(f"    First param: {category_name} = {param.get('value', 'N/A')}")
            else:
                print(f"    ⚠️ NO PARAMETERS")
        else:
            print("⚠️ NO PROCESSES")
    else:
        print("⚠️ NO ASSAYS")


STUDY 1: Case 1
Assay: a_st01_se01
Total processes: 34

First Process:
  Protocol: unknown
  Parameters: 9
    First param: N/A = Table

STUDY 8: Case 8
Assay: a_st08_se01
Total processes: 12

First Process:
  Protocol: unknown
  Parameters: 9
    First param: N/A = Table

STUDY 16: Case 16
Assay: a_st16_se01
Total processes: 12

First Process:
  Protocol: unknown
  Parameters: 9
    First param: N/A = Table


## Study Variables and Factor Values

In [32]:
# Check study factors (variables) and their values in samples
for i, study in enumerate(data["studies"][:5], 1):  # First 5 studies
    print(f"\n{'='*70}")
    print(f"STUDY {i}: {study['title']}")
    print(f"{'='*70}")
    
    factors = study.get('factors', [])
    print(f"Number of Study Factors: {len(factors)}")
    
    # Create a lookup map for factor IDs to factor names
    factor_id_to_name = {}
    for factor in factors:
        factor_id = factor.get('@id', '')
        factor_name = factor.get('factorName', 'unknown')
        if factor_id:
            factor_id_to_name[factor_id] = factor_name
    
    # List all factors
    for factor in factors:
        factor_name = factor.get('factorName', 'unknown')
        factor_type = factor.get('factorType', {}).get('annotationValue', 'unknown')
        print(f"  • {factor_name} (Type: {factor_type})")
    
    # Check factor values in samples
    samples = study.get('materials', {}).get('samples', [])
    if samples:
        print(f"\nSample count: {len(samples)}")
        # Look at first sample's factor values
        first_sample = samples[0]
        factor_values = first_sample.get('factorValues', [])
        print(f"Factor values in first sample: {len(factor_values)}")
        
        for fv in factor_values[:5]:  # Show first 5
            # Resolve factor name from reference
            category = fv.get('category', {})
            factor_id = category.get('@id', '')
            factor_ref = factor_id_to_name.get(factor_id, 'unknown')
            
            value = fv.get('value', 'N/A')
            unit = fv.get('unit', {})
            unit_term = unit.get('annotationValue', '') if unit else ''
            # Truncate long values (like filenames)
            value_str = str(value)
            if len(value_str) > 50:
                value_display = value_str[:47] + '...'
            else:
                value_display = value_str
            
            print(f"    {factor_ref}: {value_display} {unit_term}".strip())    

    else:        print("\n⚠️ NO SAMPLES")


STUDY 1: Case 1
Number of Study Factors: 6
  • Time (Type: Time)
  • Cutting Speed (Type: Operating condition)
  • Depth of Cut (Type: Operating condition)
  • Feed (Type: Operating condition)
  • Material (Type: Operating condition)
  • VB (Type: Damage)

Sample count: 17
Factor values in first sample: 6
Time: D:\DPL\ISA3\Example
  Implementations\NASA Tool...
Cutting Speed: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
Depth of Cut: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
Feed: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
Material: D:\DPL\ISA3\Example Implementations\NASA
  Tool...

STUDY 2: Case 2
Number of Study Factors: 6
  • Time (Type: Time)
  • Cutting Speed (Type: Operating condition)
  • Depth of Cut (Type: Operating condition)
  • Feed (Type: Operating condition)
  • Material (Type: Operating condition)
  • VB (Type: Damage)

Sample count: 14
Factor values in first sample: 6
Time: D:\DPL\ISA3\Example
  Implementations\NASA Tool...
Cutting Speed: D:\DPL

In [33]:
# Compare factor values across different samples (to see variation)
for i, study in enumerate(data["studies"][:3], 1):  # First 3 studies
    print(f"\n{'='*70}")
    print(f"STUDY {i}: {study['title']}")
    print(f"{'='*70}")
    
    samples = study.get('materials', {}).get('samples', [])
    factors = study.get('factors', [])
    
    if samples and factors:
        # Create a lookup map for factor IDs to factor names
        factor_id_to_name = {}
        for factor in factors:
            factor_id = factor.get('@id', '')
            factor_name = factor.get('factorName', 'unknown')
            if factor_id:
                factor_id_to_name[factor_id] = factor_name
        
        print(f"Samples: {len(samples)}, Factors: {len(factors)}")
        print("\nFactor values across first 3 samples:")
        
        for sample_idx, sample in enumerate(samples[:3], 1):
            print(f"\n  Sample {sample_idx} ({sample.get('name', 'unnamed')}):")
            factor_values = sample.get('factorValues', [])
            
            for fv in factor_values:
                # Resolve factor name from reference
                category = fv.get('category', {})
                factor_id = category.get('@id', '')
                factor_name = factor_id_to_name.get(factor_id, 'unknown')
                
                value = fv.get('value', 'N/A')
                unit = fv.get('unit', {})
                unit_term = unit.get('annotationValue', '') if unit else ''
                # Truncate long values (like filenames)
                value_str = str(value)
                if len(value_str) > 50:
                    value_display = value_str[:47] + '...'
                else:
                    value_display = value_str
                
                print(f"    • {factor_name}: {value_display} {unit_term}".strip())    

    else:        print("⚠️ NO SAMPLES OR FACTORS")


STUDY 1: Case 1
Samples: 17, Factors: 6

Factor values across first 3 samples:

  Sample 1 (Test Setup Characteristics-0):
• Time: D:\DPL\ISA3\Example
  Implementations\NASA Tool...
• Cutting Speed: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
• Depth of Cut: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
• Feed: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
• Material: D:\DPL\ISA3\Example Implementations\NASA
  Tool...
• VB: D:\DPL\ISA3\Example Implementations\NASA
  Tool...

  Sample 2 (Test Setup Characteristics-1):
• Time: D:\DPL\ISA3\Example
  Implementations\NASA Tool...
• Cutting Speed: D:\DPL\ISA3\Example Implementations\NASA Tool W...
• Depth of Cut: D:\DPL\ISA3\Example Implementations\NASA Tool W...
• Feed: D:\DPL\ISA3\Example Implementations\NASA Tool W...
• Material: D:\DPL\ISA3\Example Implementations\NASA Tool W...
• VB: D:\DPL\ISA3\Example Implementations\NASA Tool W...

  Sample 3 (Test Setup Characteristics-2):
• Time: D:\DPL\ISA3\Example
  Implementati

In [34]:
# Summary: Factor value completeness across all studies
print("Factor Value Summary Across All Studies:")
print("=" * 70)

for i, study in enumerate(data["studies"], 1):
    samples = study.get('materials', {}).get('samples', [])
    factors = study.get('factors', [])
    
    if samples:
        # Count how many factor values each sample has
        factor_value_counts = [len(s.get('factorValues', [])) for s in samples]
        avg_fv = sum(factor_value_counts) / len(factor_value_counts) if factor_value_counts else 0
        
        status = "✅" if avg_fv > 0 else "❌"
        print(f"Study {i:2d} ({study['title']:10s}): {len(factors)} factors, {len(samples):2d} samples, avg {avg_fv:.1f} FVs/sample {status}")
    else:
        print(f"Study {i:2d} ({study['title']:10s}): {len(factors)} factors, NO SAMPLES ❌")

Factor Value Summary Across All Studies:
Study  1 (Case 1    ): 6 factors, 17 samples, avg 6.0 FVs/sample ✅
Study  2 (Case 2    ): 6 factors, 14 samples, avg 6.0 FVs/sample ✅
Study  3 (Case 3    ): 6 factors, 14 samples, avg 6.0 FVs/sample ✅
Study  4 (Case 4    ): 6 factors,  7 samples, avg 6.0 FVs/sample ✅
Study  5 (Case 5    ): 6 factors,  6 samples, avg 6.0 FVs/sample ✅
Study  6 (Case 6    ): 6 factors,  1 samples, avg 6.0 FVs/sample ✅
Study  7 (Case 7    ): 6 factors,  8 samples, avg 6.0 FVs/sample ✅
Study  8 (Case 8    ): 6 factors,  6 samples, avg 6.0 FVs/sample ✅
Study  9 (Case 9    ): 6 factors,  9 samples, avg 6.0 FVs/sample ✅
Study 10 (Case 10   ): 6 factors, 10 samples, avg 6.0 FVs/sample ✅
Study 11 (Case 11   ): 6 factors, 23 samples, avg 6.0 FVs/sample ✅
Study 12 (Case 12   ): 6 factors, 15 samples, avg 6.0 FVs/sample ✅
Study 13 (Case 13   ): 6 factors, 15 samples, avg 6.0 FVs/sample ✅
Study 14 (Case 14   ): 6 factors,  9 samples, avg 6.0 FVs/sample ✅
Study 15 (Case 15   )

In [35]:
# Detailed view: Study variables per run (sample) with values
for i, study in enumerate(data["studies"][:2], 1):  # First 2 studies
    print(f"\n{'='*80}")
    print(f"STUDY {i}: {study['title']}")
    print(f"{'='*80}")
    
    factors = study.get('factors', [])
    samples = study.get('materials', {}).get('samples', [])
    
    # Create a lookup map for factor IDs to factor names
    factor_id_to_name = {}
    for factor in factors:
        factor_id = factor.get('@id', '')
        factor_name = factor.get('factorName', 'unknown')
        if factor_id:
            factor_id_to_name[factor_id] = factor_name
    
    print(f"Study Factors: {[f.get('factorName', 'unknown') for f in factors]}")
    print(f"Total Runs (Samples): {len(samples)}")
    print()
    
    # Show each run with its factor values
    for run_idx, sample in enumerate(samples[:5], 1):  # Show first 5 runs
        print(f"  Run {run_idx}: {sample.get('name', 'unnamed')}")
        factor_values = sample.get('factorValues', [])
        
        if factor_values:
            for fv in factor_values:
                # Resolve factor name from reference
                category = fv.get('category', {})
                factor_id = category.get('@id', '')
                factor_name = factor_id_to_name.get(factor_id, 'unknown')
                
                value = fv.get('value', 'N/A')
                unit = fv.get('unit', {})
                unit_term = unit.get('annotationValue', '') if unit else ''
                
                # Handle long values (filenames)
                value_str = str(value)
                if len(value_str) > 60:
                    # Extract just the filename if it's a path
                    if '\\' in value_str or '/' in value_str:
                        filename = value_str.split('\\')[-1].split('/')[-1]
                        value_display = f".../{filename}"
                    else:
                        value_display = value_str[:57] + '...'
                else:
                    value_display = value_str
                
                print(f"    • {factor_name}: {value_display} {unit_term}".strip())
        else:
            print(f"    ⚠️ NO FACTOR VALUES")
        print()


STUDY 1: Case 1
Study Factors: ['Time', 'Cutting Speed', 'Depth of Cut', 'Feed', 'Material', 'VB']
Total Runs (Samples): 17

  Run 1: Test Setup Characteristics-0
• Time: .../Case_01_time_run_01.csv
• Cutting Speed: .../Case_01_time_run_01.csv
• Depth of Cut: .../Case_01_time_run_01.csv
• Feed: .../Case_01_time_run_01.csv
• Material: .../Case_01_time_run_01.csv
• VB: .../Case_01_time_run_01.csv

  Run 2: Test Setup Characteristics-1
• Time: .../Case_01_time_run_02.csv
• Cutting Speed: .../Case_01_time_run_02.csv
• Depth of Cut: .../Case_01_time_run_02.csv
• Feed: .../Case_01_time_run_02.csv
• Material: .../Case_01_time_run_02.csv
• VB: .../Case_01_time_run_02.csv

  Run 3: Test Setup Characteristics-2
• Time: .../Case_01_time_run_03.csv
• Cutting Speed: .../Case_01_time_run_03.csv
• Depth of Cut: .../Case_01_time_run_03.csv
• Feed: .../Case_01_time_run_03.csv
• Material: .../Case_01_time_run_03.csv
• VB: .../Case_01_time_run_03.csv

  Run 4: Test Setup Characteristics-3
• Time: .../Ca

In [36]:
# Create a table showing all factor values for all runs in a study
import pandas as pd

for study_idx in [0, 1]:  # Studies 1 and 2
    study = data["studies"][study_idx]
    print(f"\n{'='*80}")
    print(f"STUDY {study_idx + 1}: {study['title']} - Factor Values Table")
    print(f"{'='*80}\n")
    
    factors = study.get('factors', [])
    samples = study.get('materials', {}).get('samples', [])
    
    if not samples or not factors:
        print("⚠️ NO SAMPLES OR FACTORS")
        continue
    
    # Create a lookup map for factor IDs to factor names
    factor_id_to_name = {}
    for factor in factors:
        factor_id = factor.get('@id', '')
        factor_name = factor.get('factorName', 'unknown')
        if factor_id:
            factor_id_to_name[factor_id] = factor_name
    
    # Build a table with runs as rows and factors as columns
    table_data = []
    
    for run_idx, sample in enumerate(samples, 1):
        row = {'Run': run_idx, 'Sample_Name': sample.get('name', 'unnamed')}
        factor_values = sample.get('factorValues', [])
        
        # Create a dict of factor name -> value for easy lookup
        fv_dict = {}
        for fv in factor_values:
            # Resolve factor name from reference
            category = fv.get('category', {})
            factor_id = category.get('@id', '')
            factor_name = factor_id_to_name.get(factor_id, 'unknown')
            
            value = fv.get('value', 'N/A')
            # Truncate long values
            value_str = str(value)
            if len(value_str) > 30:
                if '\\' in value_str or '/' in value_str:
                    value_str = '.../' + value_str.split('\\')[-1].split('/')[-1]
                    if len(value_str) > 30:
                        value_str = value_str[:27] + '...'
                else:
                    value_str = value_str[:27] + '...'
            fv_dict[factor_name] = value_str
        
        # Add all factor values to row
        for factor in factors:
            factor_name = factor.get('factorName', 'unknown')
            row[factor_name] = fv_dict.get(factor_name, 'N/A')
        
        table_data.append(row)
    
    # Create and display DataFrame
    df = pd.DataFrame(table_data)
    print(df.to_string(index=False, max_colwidth=30))
    print()


STUDY 1: Case 1 - Factor Values Table

 Run                   Sample_Name                        Time               Cutting Speed                Depth of Cut                        Feed                    Material                          VB
   1  Test Setup Characteristics-0 .../Case_01_time_run_01.csv .../Case_01_time_run_01.csv .../Case_01_time_run_01.csv .../Case_01_time_run_01.csv .../Case_01_time_run_01.csv .../Case_01_time_run_01.csv
   2  Test Setup Characteristics-1 .../Case_01_time_run_02.csv .../Case_01_time_run_02.csv .../Case_01_time_run_02.csv .../Case_01_time_run_02.csv .../Case_01_time_run_02.csv .../Case_01_time_run_02.csv
   3  Test Setup Characteristics-2 .../Case_01_time_run_03.csv .../Case_01_time_run_03.csv .../Case_01_time_run_03.csv .../Case_01_time_run_03.csv .../Case_01_time_run_03.csv .../Case_01_time_run_03.csv
   4  Test Setup Characteristics-3 .../Case_01_time_run_04.csv .../Case_01_time_run_04.csv .../Case_01_time_run_04.csv .../Case_01_time_run_04.csv .

In [37]:
# Debug: Check the actual factor_name structure in the ISA JSON
study1 = data["studies"][0]
sample1 = study1["materials"]["samples"][0]

print("Study Factors (definitions):")
for factor in study1["factors"]:
    print(f"  • Name: {factor.get('factorName')}, Type: {factor.get('factorType', {}).get('annotationValue')}")

print("\n" + "="*70)
print("Factor Values in First Sample:")
print("="*70)
for fv in sample1.get("factorValues", []):
    print(f"\nFactor Value:")
    print(f"  category: {fv.get('category')}")
    print(f"  value: {str(fv.get('value'))[:60]}...")
    
    # Check if category has factorName
    category = fv.get('category', {})
    if isinstance(category, dict):
        print(f"  category.factorName: {category.get('factorName', 'MISSING')}")
        print(f"  category keys: {list(category.keys())}")

Study Factors (definitions):
  • Name: Time, Type: Time
  • Name: Cutting Speed, Type: Operating condition
  • Name: Depth of Cut, Type: Operating condition
  • Name: Feed, Type: Operating condition
  • Name: Material, Type: Operating condition
  • Name: VB, Type: Damage

Factor Values in First Sample:

Factor Value:
  category: {'@id': '#study_factor/269e20bb-2fbe-44a1-8ed4-5a87bca7e940'}
  value: D:\DPL\ISA3\Example
  Implementations\NASA Tool Wear\01)
  N...
  category.factorName: MISSING
  category keys: ['@id']

Factor Value:
  category: {'@id': '#study_factor/572e1df1-9555-48f1-bf08-04e9da40f36c'}
  value: D:\DPL\ISA3\Example Implementations\NASA
  Tool Wear\01)
  N...
  category.factorName: MISSING
  category keys: ['@id']

Factor Value:
  category: {'@id': '#study_factor/e93752a8-bae0-4e7f-94cb-8ba1e3d2d38c'}
  value: D:\DPL\ISA3\Example Implementations\NASA
  Tool Wear\01)
  N...
  category.factorName: MISSING
  category keys: ['@id']

Factor Value:
  category: {'@id': '#study

## ✅ SOLUTION: Factor Name Resolution

### The "N/A" Problem Explained

The ISA-JSON format uses **references** instead of embedding full objects. When the Python ISA tools serialize the data:

**What we expected:**
```json
{
  "category": {
    "factorName": "Time",
    "@id": "#study_factor/abc123"
  }
}
```

**What ISA-JSON actually produces:**
```json
{
  "category": {
    "@id": "#study_factor/abc123"
  }
}
```

The full factor details (like `factorName`) are stored separately in the `study.factors` array, and factor values reference them by `@id`.

### The Fix

To get the factor names, we need to:
1. Build a lookup map: `factor_id → factor_name`
2. When reading factor values, resolve the `@id` reference using this map

```python
# Create lookup
factor_id_to_name = {}
for factor in study['factors']:
    factor_id_to_name[factor['@id']] = factor['factorName']

# Resolve reference
factor_id = fv['category']['@id']
factor_name = factor_id_to_name[factor_id]
```

### Results After Fix
✅ Factor names now display correctly (Time, Cutting Speed, etc.)  
✅ Each run has correct per-run values (run_01, run_02, run_03...)  
✅ Table now shows actual values instead of N/A

In [38]:
# Verify the bug: Check if app-out.json has different values per run
with open("app-out.json", "r") as f:
    app_data = json.load(f)

study1 = app_data["studies"][0]
mappings = study1["study_to_study_variable_mapping"]

print(f"Total mappings for Study 1: {len(mappings)}")
print(f"Expected: {study1['total_runs']} runs × 6 variables = {study1['total_runs'] * 6} mappings")
print()

# Show Time variable mappings for first 5 runs
print("Time variable values for first 5 runs:")
time_mappings = [m for m in mappings if m.get("variableName") == "Time"][:5]
for m in time_mappings:
    run_num = m.get("runNumber")
    value = m.get("value", "")
    filename = value.split('\\')[-1].split('/')[-1]
    print(f"  Run {run_num}: {filename}")

Total mappings for Study 1: 102
Expected: 17 runs × 6 variables = 102 mappings

Time variable values for first 5 runs:
  Run 1: Case_01_time_run_01.csv
  Run 2: Case_01_time_run_02.csv
  Run 3: Case_01_time_run_03.csv
  Run 4: Case_01_time_run_04.csv
  Run 5: Case_01_time_run_05.csv


## ✅ Confirmed: The Bug

**app-out.json HAS the correct data:**
- Run 1 → `time_run_01.csv`
- Run 2 → `time_run_02.csv`
- Run 3 → `time_run_03.csv`
- etc.

**But web-to-isa-phm.py ONLY uses the first mapping** (line 347):
```python
mapping = mapping[0]  # ⚠️ Always gets run-01
```

---

## 🔧 The Fix (Pseudo-code)

The correct approach should be:

```python
# WRONG (current approach):
# 1. Create ONE dummy sample with factor values from run 1
# 2. Clone it 17 times → all get run 1 values

# CORRECT (needed approach):
# 1. Create a base dummy sample with characteristics (no factor values yet)
# 2. Clone it 17 times to get 17 samples
# 3. Loop through each sample and assign factor values specific to that run:

for run_number in range(1, study_total_runs + 1):
    sample = study_obj.samples[run_number - 1]
    
    for variable in study_variables:
        # Find mapping for THIS specific run AND variable
        mapping = find_mapping_where(
            variableName == variable.name AND
            runNumber == run_number
        )
        
        factor_value = FactorValue()
        factor_value.factor_name = study_factor_obj
        factor_value.value = mapping.get("value")
        factor_value.unit = ...
        
        sample.factor_values.append(factor_value)
```

This ensures each run gets its correct filenames.

## 🔧 Fix Applied to web-to-isa-phm.py

The conversion script has been updated with the following changes:

### What Changed:
1. **Separated factor creation from value assignment**
   - First: Create all study factors (definitions only)
   - Then: Create samples using `batch_create_materials()`
   - Finally: Loop through samples and assign factor values per run

2. **Fixed the mapping lookup**
   ```python
   # OLD (buggy):
   mapping = list(filter(lambda x: x.get("variableName") == study_factor_obj.name, ...))
   mapping = mapping[0]  # Always first = run 1
   
   # NEW (fixed):
   mapping = next(
       (m for m in study.get("study_to_study_variable_mapping", [])
        if m.get("variableName") == variable_name AND m.get("runNumber") == run_number),
       None
   )
   ```

3. **Now iterates over all runs**
   ```python
   for run_number in range(1, study_total_runs + 1):
       sample = study_obj.samples[run_number - 1]
       # Assign factor values specific to this run
   ```

### Result:
- ✅ Run 1 gets `time_run_01.csv`
- ✅ Run 2 gets `time_run_02.csv`
- ✅ Run 5 gets `time_run_05.csv`
- ✅ Factor names will be correct (not "unknown")

## 🐛 Bug Analysis: Factor Values Issues

### Problem 1: Factor names showing as "unknown"
**Root Cause:** Lines 344-352 in `web-to-isa-phm.py`

The code creates ONE `FactorValue` per study variable and assigns it to the `dummy_sample`. It then uses `batch_create_materials()` to clone this dummy sample 17 times. However, when cloning, the `factor_value.factor_name` reference is lost or not properly copied.

```python
# Current buggy code:
factor_value = FactorValue()
factor_value.factor_name = study_factor_obj  # This is a reference
mapping = list(filter(lambda x: x.get("variableName") == study_factor_obj.name, ...))
if mapping:
    mapping = mapping[0]  # ⚠️ BUG: Only gets FIRST mapping (run 1)
    factor_value.value = mapping.get("value", "unknown")
```

### Problem 2: All runs get the same filename (run_01)
**Root Cause:** Same code block - Line 347

The filter finds ALL mappings for a variable name (e.g., all "Time" entries across all runs), but then it only takes the FIRST one (`mapping[0]`). This means:
- Run 1 gets `Case_01_time_run_01.csv` ✅
- Run 2 gets `Case_01_time_run_01.csv` ❌ (should be run_02)
- Run 3 gets `Case_01_time_run_01.csv` ❌ (should be run_03)
- etc.

### The Data Structure in app-out.json
Each study has `study_to_study_variable_mapping` array with entries like:
```json
{
  "studyRunId": "..::run-01",
  "runNumber": 1,
  "studyVariableId": "...",
  "value": "...time_run_01.csv",
  "variableName": "Time"
}
```

There are 17 runs × 6 variables = 102 mapping entries for Study 1, but the code only uses the first 6 (all run-01).

### Solution Required
The code needs to:
1. Loop through each run number (1 to total_runs)
2. For each run, create a sample
3. For each sample, assign factor values specific to that run number
4. Match mappings by BOTH `variableName` AND `runNumber`

In [39]:
# Extract all measurement protocol parameters across all studies
import pandas as pd

measurement_data = []

for study_idx, study in enumerate(data["studies"], 1):
    for assay in study.get('assays', []):
        for process in assay.get('processSequence', []):
            protocol = process.get('executesProtocol', {})
            protocol_name = protocol.get('name', 'unknown')
            protocol_type = protocol.get('protocolType', {}).get('annotationValue', '')
            
            # Only process measurement protocols
            if 'measurement' in protocol_name.lower():
                for param_val in process.get('parameterValues', []):
                    category = param_val.get('category', {})
                    param_name = category.get('parameterName', {}).get('annotationValue', 'unknown')
                    value = param_val.get('value', 'N/A')
                    unit = param_val.get('unit', {})
                    unit_term = unit.get('annotationValue', '') if unit else ''
                    
                    measurement_data.append({
                        'Study': f"Study {study_idx}",
                        'Protocol': protocol_name,
                        'Parameter': param_name,
                        'Value': value,
                        'Unit': unit_term
                    })

# Extract all processing protocol parameters across all studies
processing_data = []

for study_idx, study in enumerate(data["studies"], 1):
    for assay in study.get('assays', []):
        for process in assay.get('processSequence', []):
            protocol = process.get('executesProtocol', {})
            protocol_name = protocol.get('name', 'unknown')
            protocol_type = protocol.get('protocolType', {}).get('annotationValue', '')
            
            # Only process processing protocols
            if 'processing' in protocol_name.lower():
                for param_val in process.get('parameterValues', []):
                    category = param_val.get('category', {})
                    param_name = category.get('parameterName', {}).get('annotationValue', 'unknown')
                    value = param_val.get('value', 'N/A')
                    unit = param_val.get('unit', {})
                    unit_term = unit.get('annotationValue', '') if unit else ''
                    
                    processing_data.append({
                        'Study': f"Study {study_idx}",
                        'Protocol': protocol_name,
                        'Parameter': param_name,
                        'Value': value,
                        'Unit': unit_term
                    })

# Create a consolidated view: unique parameter-value-unit combinations
print("=" * 100)
print("MEASUREMENT PROTOCOLS - Unique Parameter Values Summary")
print("=" * 100)

if measurement_data:
    df_m = pd.DataFrame(measurement_data)
    
    # Group by parameter and show unique values
    for param_name in sorted(df_m['Parameter'].unique()):
        param_rows = df_m[df_m['Parameter'] == param_name]
        unique_values = param_rows[['Value', 'Unit']].drop_duplicates()
        
        print(f"\n{param_name}:")
        if len(unique_values) <= 10:
            for _, row in unique_values.iterrows():
                val = row['Value']
                unit = row['Unit']
                print(f"  • {val} {unit}".strip())
        else:
            print(f"  ({len(unique_values)} unique values)")
            for _, row in unique_values.head(5).iterrows():
                val = row['Value']
                unit = row['Unit']
                print(f"  • {val} {unit}".strip())
            print(f"  ... and {len(unique_values) - 5} more")
else:
    print("No measurement protocol parameters found")

print("\n\n" + "=" * 100)
print("PROCESSING PROTOCOLS - Unique Parameter Values Summary")
print("=" * 100)

if processing_data:
    df_p = pd.DataFrame(processing_data)
    
    # Group by parameter and show unique values
    for param_name in sorted(df_p['Parameter'].unique()):
        param_rows = df_p[df_p['Parameter'] == param_name]
        unique_values = param_rows[['Value', 'Unit']].drop_duplicates()
        
        print(f"\n{param_name}:")
        if len(unique_values) <= 10:
            for _, row in unique_values.iterrows():
                val = row['Value']
                unit = row['Unit']
                print(f"  • {val} {unit}".strip())
        else:
            print(f"  ({len(unique_values)} unique values)")
            for _, row in unique_values.head(5).iterrows():
                val = row['Value']
                unit = row['Unit']
                print(f"  • {val} {unit}".strip())
            print(f"  ... and {len(unique_values) - 5} more")
else:
    print("No processing protocol parameters found")

MEASUREMENT PROTOCOLS - Unique Parameter Values Summary
No measurement protocol parameters found


PROCESSING PROTOCOLS - Unique Parameter Values Summary
No processing protocol parameters found


## Measurement & Processing Protocol Parameters

## Protocol Parameter Grids (Protocol × Sensor Mappings) using isa